File: java/src/main/java/TopologicalSort.java
import java.util.*;

public class TopologicalSort {

	static void dfs(List<Integer>[] graph, boolean[] used, List<Integer> res, int u) {
		used[u] = true;
		for (int v : graph[u])
			if (!used[v])
				dfs(graph, used, res, v);
		res.add(u);
	}

	public static List<Integer> topologicalSort(List<Integer>[] graph) {
		int n = graph.length;
		boolean[] used = new boolean[n];
		List<Integer> res = new ArrayList<>();
		for (int i = 0; i < n; i++)
			if (!used[i])
				dfs(graph, used, res, i);
		Collections.reverse(res);
		return res;
	}

	// Usage example
	public static void main(String[] args) {
		int n = 3;
		List<Integer>[] g = new List[n];
		for (int i = 0; i < n; i++) {
			g[i] = new ArrayList<>();
		}
		g[2].add(0);
		g[2].add(1);
		g[0].add(1);

		List<Integer> res = topologicalSort(g);
		System.out.println(res);
	}
}

File: python/topological_sort.py
# https://en.wikipedia.org/wiki/Topological_sorting


def topological_sort(graph):
    def dfs(graph, used, order, u):
        used[u] = True
        for v in graph[u]:
            if not used[v]:
                dfs(graph, used, order, v)
        order.append(u)

    n = len(graph)
    used = [False] * n
    order = []
    for i in range(n):
        if not used[i]:
            dfs(graph, used, order, i)

    return order[::-1]


def test():
    g = [[] for _ in range(3)]
    g[2].append(0)
    g[2].append(1)
    g[0].append(1)

    assert topological_sort(g) == [2, 0, 1]


test()
